### ISSUE TITLE
Support input types in CS

### ISSUE THREAD

**Issue by Bonczidai** [CloudSlang/cloud-slang#673]
**[APPROACH I - STRINGS]** go with string interface for now, support types later

**Comment by Bonczidai** [comment]
Functional spec:
- for now we will only support string type in CloudSlang for executable interface
- verification is done at runtime
- executable input / output need to be resolved to string after binding (no implicit casting)
- step input / publish value need to be resolved to string after binding (no implicit casting)
- action outputs are available in operation outputs and results section as objects (no change)
- loops: TBD

p.s. executable = flow / op

**Comment by Bonczidai** [comment]
examples:

```yaml
operation:
  name: print

  inputs:
    - input_1 # fails if not string
    - input_2: '123'
    - input_3: 123 # fails

  action:
    python_script: |
      val_int = 123
      val_str = 'abc'
      val_bool = True

  outputs:
    - output_1 # fails if output_1 is not string
    - output_2: 'hello'
    - output_3: val_bool # fails
    - output_4: str(val_bool)

  results:
    - SUCCESS: ${text != 'fail'}
    - FAILURE
```

```yaml
flow:
  name: print_flow

  workflow:
    - step_name:
        do:
          executable_name:
            - input_1 # fails if not string
            - input_2: 123 # fails
            - input_3: '123'
        publish:
          - publish_int: 123 # fails
          - publish_str: 'abc'
```

**Comment by Bonczidai** [comment]
for loop example:

```yaml
flow:
  name: print_flow
  inputs:
    - list_as_str: '1 2 3'

  workflow:

  - print_message:
      loop:
        for: nr in list_as_str.split() # `list_as_str` also works
        do:
          print:
            - text: ${nr}
```

basically we need a delimiter and convert vice-versa between string and array representations (e.g. join(), split() functions)

**Comment by orius123** [comment]
Y

**Comment by Bonczidai** [comment]
**[APPROACH II - TYPES]**

- Current supported types: String \ Number \ Boolean \ Object (this will also give us in the future ability to support more complex types like dates)
- Default type (if not specified) is String 
- we want to validate the type during the binding (runtime - e.g. at the beginning of the flow \ operation)
- no implicit type conversion: 123 for String input fails, '123' passes
- if type is not correct -> illegal argument – type mismatch message
- during the flow the type might change (e.g. using the publish section) – we do not validate it!
- we will validate primitive default value is of specified type (compile time):
```
- input_key:
    default: 123 # fails
    type: string
```

examples:

```yaml
operation:
  name: print

  inputs:
    - input_1 # type: string
    - input_2: '123'
    - input_3: 123 # fails
    - input_4:
         default: 123
         type: Number
    - input_5:
         default: [1,2,3]
         type: Object

  action:
    python_script: |
      val_int = 123
      val_str = 'abc'
      val_bool = True

  outputs:
    - output_1 # fails if output_1 is not string
    - output_2: 'hello'
    - output_3: 
         default: val_bool
         type: Boolean

  results:
    - SUCCESS: ${val_str != 'fail'}
    - FAILURE
```

@CloudSlang/slang-functional-reviewers WDYT?

**Comment by lucian-cm** [comment]
Number seems to be a big concept. Currently there isn't one representation that is good for both floating point and integer-like numbers. So I guess the question is how is it planned to represent numbers in the back-end and what exactly is a number? 
If we plan to support floating point we should add more details on what formats default values can be given in like scientific notation, decimal point, locale handling.
Some strongly typed languages allow signed and unsigned integers. It might be important to tell the binary representation scheme (2's complement) for internal representation and endianess in case of serialization.

**Comment by lucian-cm** [comment]
Type changing at runtime is a powerfull construct, but can be in itself something hard to follow when flows get bigger spanning many files, might be useful to have a sort of safe net and have a way to infer the type at runtime at any given time using a sort of getType or 'instanceof' trick or to assert types.

**Comment by lucian-cm** [comment]
Mostly types exist to categorize items of the same kind and because they have a commons of operations that they support. What are the operations particular to each type are they only operators like + - / * , are they only python expressions?

**Comment by lucian-cm** [comment]
Is there plan to support explicit type conversion between types? There might be use-cases when that is useful.

**Comment by orius123** [comment]
what is the motivation?

**Comment by Bonczidai** [comment]
yupp - there are concerns on how to implement it correctly, so we will go with `Approach I` for now
